### PR TITLE
feat: implement udf server in databend

### LINKED ISSUE #12177: Tracking: Implement Databend UDF Server
This issue is used to tracking the progress of implementing Databend UDF Server.
**RFC:** #12075

## Tasks
- [x] #12168
- [x] #12417
- [x] #12463 
- [x] #12463 
- [ ] #12802

### PR DESCRIPTION
**Pull request by gitccl**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary
This pr merges previous #12168  , #12417 and #12463 .
In this pr, we also change the syntax of `create function` command to make it similar to snowflake's syntax.
```
CREATE FUNCTION [IF NOT EXISTS] <udf_name> (<arg_type>, ...) RETURNS <return_type> LANGUAGE <language> HANDLER=<handler> ADDRESS=<udf_server_address>
```

- Closes #12177

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12729)

### CODE REVIEW

**Review by drmingdrmer — COMMENTED** [review]
> Reviewed 3 of 71 files at r1, 8 of 57 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12729)* status: 11 of 72 files reviewed, all discussions resolved

**Review by drmingdrmer — APPROVED** [review]

**Comment by xudong963** [comment]
I'll review the PR today

File: .github/workflows/reuse.linux.yml [hunk 1]
```diff
@@ -189,6 +189,35 @@ jobs:
         with:
           name: test-sqllogic-standalone-${{ matrix.handlers }}-${{ matrix.dirs }}
 
+  sqllogic_standalone_udf_server:
```
Comment (xudong963) [inline .github/workflows/reuse.linux.yml:192]: Ping @everpcpc to check the ci.

**Review by xudong963 — APPROVED** [review]
LGTM, great work @gitccl 

Currently, our Python UDF syntax is similar to snowflake:

Snowflake:
```
create or replace function addone(i int)
returns int
language python
runtime_version = '3.8'
handler = 'addone_py'
as
$$
def addone_py(i):
  return i+1
$$;
```

Databend:
```
CREATE FUNCTION split_and_join (VARCHAR, VARCHAR, VARCHAR) RETURNS VARCHAR LANGUAGE python HANDLER = 'split_and_join' ADDRESS = 'http://0.0.0.0:8815';
```

@BohuTANG @sundy-li  Any thoughts?

**Comment by xudong963** [comment]
Given the security concerns we discussed internally, maybe we still need to add a flag to make it a feature, so we can avoid our cloud attacking. cc @Xuanwo @flaneur2020 @BohuTANG

**Comment by xudong963** [comment]
cc @soyeric128 After the PR merged, need to add a doc. You can refer the README:  https://github.com/datafuselabs/databend/pull/12729/files#diff-bf3534b9c183942327d5e8579cec90a1634bc242312a7cff036d80c035d11d98

File: .github/workflows/reuse.linux.yml (hunk 1 above)
Comment (everpcpc) [reply .github/workflows/reuse.linux.yml:192]: CI failed:
https://github.com/datafuselabs/databend/actions/runs/6098657496/job/16549342052?pr=12729

**Comment by xudong963** [comment]
Seems the ee test  isn't related to the PR

image

**Comment by BohuTANG** [comment]
> Given the security concerns we discussed internally, maybe we still need to add a flag to make it a feature, so we can avoid our cloud attacking. cc @Xuanwo @flaneur2020 @BohuTANG

Add a new query config is ok, then we can enable it for some users. @flaneur2020

File: .github/workflows/reuse.linux.yml (hunk 1 above)
Comment (gitccl) [reply .github/workflows/reuse.linux.yml:192]: fixed

**Comment by gitccl** [comment]
It seems that ci is stuck.

**Comment by Xuanwo** [comment]
> It seems that ci is stuck.

Please fix the conflicts first to make CI running, thanks!

**Comment by xudong963** [comment]
image

A flaky test cc @TCeason

**Comment by TCeason** [comment]
> image
> 
> A flaky test cc @TCeason

Get

File: src/query/expression/src/evaluator.rs [hunk 2]
```diff
@@ -195,6 +208,100 @@ impl<'a> Evaluator<'a> {
...
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let input_batch = DataBlock::new(block_entries, num_rows)
+            .to_record_batch_keep_schema(&data_schema)
+            .map_err(|err| ErrorCode::from_string(format!("{err}")))?;
+
+        let func_name = func_name.to_string();
+        let server_addr = server_addr.to_string();
+        let result_batch = GlobalQueryRuntime::instance()
```
Comment (xudong963) [inline src/query/expression/src/evaluator.rs:262]: We should make `run_udf_server_call` async, or it's possible to make threads hang and can't be executed by batch.  Suggestions from @zhang2014

Comment (xudong963) [reply src/query/expression/src/evaluator.rs:262]: Let's do it in another PR

**Comment by BohuTANG** [comment]
Let's wait the databend.rs works again,  some tests have failed due to this problem:/

File: tests/sqllogictests/suites/udf_server/udf_server_test [hunk 3]
```diff
@@ -0,0 +1,363 @@
+# Please start the UDF Server first before running this test:
+#   python3 tests/udf-server/udf_test.py
+#
+
+statement ok
+set enable_udf_server = 1;
```
Comment (flaneur2020) [inline tests/sqllogictests/suites/udf_server/udf_server_test:6]: is this config considered as a dyanmic settings which can be enabled by any user? 

IMHO it'd be better to limit this configuration in the config file, and can not be changed in the runtime.

File: src/query/expression/src/utils/udf_client.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+#[derive(Debug, Clone)]
+pub struct UDFFlightClient {
+    inner: FlightServiceClient<Channel>,
+}
+
+impl UDFFlightClient {
+    #[async_backtrace::framed]
+    pub async fn connect(addr: &str) -> Result<UDFFlightClient> {
+        let inner = FlightServiceClient::connect(addr.to_string())
```
Comment (zhang2014) [inline src/query/expression/src/utils/udf_client.rs:44]: Need connect timeout.

File: tests/sqllogictests/suites/udf_server/udf_server_test (hunk 3 above)
Comment (gitccl) [reply tests/sqllogictests/suites/udf_server/udf_server_test]: done

File: src/query/expression/src/utils/udf_client.rs (hunk 4 above)
Comment (gitccl) [reply src/query/expression/src/utils/udf_client.rs]: done

**Comment by xudong963** [comment]
```
Databend Query start failure, cause: Code: 1002, Text = TOML parse error at line 41, column 1
553
   |
554
41 | enable_udf_server = true
555
   | ^^^^^^^^^^^^^^^^^
556
unknown field `enable_udf_server`, expected one of `tenant_id`, `cluster_id`, `num_cpus`, `mysql_handler_host`, `mysql_handler_port`, `mysql_handler_tcp_keepalive_timeout_secs`, `max_active_sessions`, `max_server_memory_usage`, `max_memory_limit_enabled`, `clickhouse_handler_host`, `clickhouse_handler_port`, `clickhouse_http_handler_host`, `clickhouse_http_handler_port`, `http_handler_host`, `http_handler_port`, `http_handler_result_timeout_secs`, `flight_sql_handler_host`, `flight_sql_handler_port`, `flight_api_address`, `admin_api_address`, `metric_api_address`, `http_handler_tls_server_cert`, `http_handler_tls_server_key`, `http_handler_tls_server_root_ca_cert`, `flight_sql_tls_server_cert`, `flight_sql_tls_server_key`, `api_tls_server_cert`, `api_tls_server_key`, `api_tls_server_root_ca_cert`, `rpc_tls_server_cert`, `rpc_tls_server_key`, `rpc_tls_query_server_root_ca_cert`, `rpc_tls_query_service_domain_name`, `table_engine_memory_enabled`, `wait_timeout_mills`, `max_query_log_size`, `management_mode`, `jwt_key_file`, `jwt_key_files`, `default_storage_format`, `default_compression`, `users`, `share_endpoint_address`, `share_endpoint_auth_token_file`, `quota`, `internal_enable_sandbox_tenant`, `internal_merge_on_read_mutation`, `table_disk_cache_mb_size`, `table_meta_cache_enabled`, `table_cache_block_meta_count`, `table_memory_cache_mb_size`, `table_disk_cache_root`, `table_cache_snapshot_count`, `table_cache_statistic_count`, `table_cache_segment_count`, `table_cache_bloom_index_meta_count`, `table_cache_bloom_index_filter_count`, `table_cache_bloom_index_data_bytes`, `disable_system_table_load`, `openai_api_key`
557
, source: None
558
TOML parse error at line 41, column 1
```
cc @gitccl

**Comment by gitccl** [comment]
> ```
> Databend Query start failure, cause: Code: 1002, Text = TOML parse error at line 41, column 1
> 553
>    |
> 554
> 41 | enable_udf_server = true
> 555
>    | ^^^^^^^^^^^^^^^^^
> 556
> unknown field `enable_udf_server`, expected one of `tenant_id`, `cluster_id`, `num_cpus`, `mysql_handler_host`, `mysql_handler_port`, `mysql_handler_tcp_keepalive_timeout_secs`, `max_active_sessions`, `max_server_memory_usage`, `max_memory_limit_enabled`, `clickhouse_handler_host`, `clickhouse_handler_port`, `clickhouse_http_handler_host`, `clickhouse_http_handler_port`, `http_handler_host`, `http_handler_port`, `http_handler_result_timeout_secs`, `flight_sql_handler_host`, `flight_sql_handler_port`, `flight_api_address`, `admin_api_address`, `metric_api_address`, `http_handler_tls_server_cert`, `http_handler_tls_server_key`, `http_handler_tls_server_root_ca_cert`, `flight_sql_tls_server_cert`, `flight_sql_tls_server_key`, `api_tls_server_cert`, `api_tls_server_key`, `api_tls_server_root_ca_cert`, `rpc_tls_server_cert`, `rpc_tls_server_key`, `rpc_tls_query_server_root_ca_cert`, `rpc_tls_query_service_domain_name`, `table_engine_memory_enabled`, `wait_timeout_mills`, `max_query_log_size`, `management_mode`, `jwt_key_file`, `jwt_key_files`, `default_storage_format`, `default_compression`, `users`, `share_endpoint_address`, `share_endpoint_auth_token_file`, `quota`, `internal_enable_sandbox_tenant`, `internal_merge_on_read_mutation`, `table_disk_cache_mb_size`, `table_meta_cache_enabled`, `table_cache_block_meta_count`, `table_memory_cache_mb_size`, `table_disk_cache_root`, `table_cache_snapshot_count`, `table_cache_statistic_count`, `table_cache_segment_count`, `table_cache_bloom_index_meta_count`, `table_cache_bloom_index_filter_count`, `table_cache_bloom_index_data_bytes`, `disable_system_table_load`, `openai_api_key`
> 557
> , source: None
> 558
> TOML parse error at line 41, column 1
> ```
> 
> cc @gitccl

This problem is caused by the 1.0.56 version of databend-query loading the latest configuration, but version 1.0.56 does not support the `enable_udf_server` configuration. Is it expected to use the old version databend-query to load the new version configuration?

**Comment by sundy-li** [comment]
> Is it expected to use the old version databend-query to load the new version configuration?

Currently, it's forbidden. Maybe we should fetch the old config file from git to start the old version process.

**Comment by sundy-li** [comment]
We already use the old config files in https://github.com/datafuselabs/databend/blob/1d19af8b180b3b88f3e9eba9837d31815c9c9364/tests/compat/test-compat.sh

**Comment by gitccl** [comment]
> We already use the old config files in https://github.com/datafuselabs/databend/blob/1d19af8b180b3b88f3e9eba9837d31815c9c9364/tests/compat/test-compat.sh


test_compat_fuse use this scripts: https://github.com/datafuselabs/databend/blob/1d19af8b180b3b88f3e9eba9837d31815c9c9364/tests/fuse-compat/util.sh

**Review by sundy-li — APPROVED** [review]

**Pull request merged**